resources/type-system/flow-analysis-v0.md

```diff
@@ -0,0 +1,339 @@
+# Flow Analysis for Non-nullability
+
+paulberry@google.com, leafp@google.com
+
+## Terminology and Notation
+
+We use the generic term *node* to denote an expression, statement, top level
+function, method, constructor, or field.  We assume that all nodes are labelled
+uniquely in some unspecified fashion such that mappings can be maintained from
+nodes to the results of flow analysis.
+
+The flow analysis assumes that the set of variables which are assigned in any
+given node has been pre-computed, and can be queried by a predicate **Assigned**
+such that for a node `N` and a variable `v`, **Assigned**(`N`, `v`) is true iff
+`N` syntactically contains an assignment to `v` (regardless of reachability of
+that assignment).
+
+### Models
+
+A *type test site* is a location in the program where a variable's type is
+tested, either via an `is` expression or a cast.
+
+A *variable model*, denoted `VariableModel(declaredType, promotedTypes,
+testSites, assigned, unassigned, writeCaptured)`, represents what is statically
+known to the flow analysis about the state of a variable at a given point in the
+source code.
+
+- `declaredType` is the type assigned to the variable at its declaration site
+  (either explicitly or by type inference).
+
+- `promotedTypes` is an ordered set of types that the variable has been promoted
+  to, with the final entry in the ordered set being the current promoted type of
+  the variable.  Note that each entry in the ordered set must be a subtype of
+  all previous entries, and of the declared type.
+
+- `testSites` is an ordered list of *type test site*s representing the set of
+  type tests that are known to have been performed on the variable in all code
+  paths leading to the given point in the source code.
+
+- `assigned` is a boolean value indicating whether the variable is known to
+  have been definitely assigned at the given point in the source code.
+
+- `unassigned` is a boolean value indicating whether the variable is known not
+  to have been definitely assigned at the given point in the source code.  (Note
+  that a variable may not be both definitely assigned and definitely
+  unassigned).
+
+- `writeCaptured` is a boolean value indicating whether a closure might exist at
+  the given point in the source code, which could potentially write to the
+  variable.
+
+A *flow model*, denoted `FlowModel(reachable, variableInfo)`, represents what
+is statically known to flow analysis about the state of the program at a given
+point in the source code.
+
+  - `reachable` is a stack of boolean values modeling the reachability of the
```

Comment:
So what's going on here?  This is my attempt to incorporate Paul's extended reachability into the basic framework.  I think this is roughly what the extended reachability was doing, except that instead of keeping the stack of previous reachability values on the call stack, I'm keeping it explicitly in the analysis.  An implementation could do either.  So  far, I've found the book-keeping a little cleaner here.  Note that I'm possibly applying this in more places than Paul had originally intended?  For example, I think the following is a motivating example:

```
Object x;
if (false && x is int) {
  x.isEven
}
```

I'm not sure whether Paul intended this to promote in the body or not, but it seemed reasonable, and I *think* if I have this set up right, then we will get that.